### PR TITLE
[core] fix incorrect conversion of valid url

### LINKED ISSUE #23282: expo-web-browser cannot handle links with more than one # in the URL on iOS
### Minimal reproducible example

https://snack.expo.dev/@gkasdorf/44ad3e

### Summary

On iOS, expo-web-browser is not able to handle links that have multiple # symbols in the URL. It's possible this occurs with other characters, but I have been unable to verify that. I am also not sure if this is an issue with all combinations or not, but I have tried varying combinations with similar results.

In some instances, such as in the Snack, the browser will simply not respond while in others, the app will crash (possibly due to a different version being used in the former case).

### Environment

  expo-env-info 1.0.5 environment info:
    System:
      OS: macOS 13.4.1
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 18.16.0 - ~/.nvm/versions/node/v18.16.0/bin/node
      Yarn: 1.22.19 - ~/.nvm/versions/node/v18.16.0/bin/yarn
      npm: 9.6.7 - ~/.nvm/versions/node/v18.16.0/bin/npm
      Watchman: 2023.07.03.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.12.1 - /opt/homebrew/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 22.4, iOS 16.4, macOS 13.3, tvOS 16.4, watchOS 9.4
    IDEs:
      Android Studio: 2022.1 AI-221.6008.13.2211.9619390
      Xcode: 14.3.1/14E300c - /usr/bin/xcodebuild
    npmPackages:
      expo: ~48.0.18 => 48.0.19 
      react: 18.2.0 => 18.2.0 
      react-dom: 18.2.0 => 18.2.0 
      react-native: 0.71.8 => 0.71.8 
      react-native-web: ~0.18.7 => 0.18.12 
    Expo Workflow: bare

### PR DESCRIPTION
**Pull request by alanjhughes**
# Why
Closes #23282 
Some unusual but valid URLs will fail to be created with `URL(string:)` 
example: https://matrix.to/#/#sh.itjust.works:matrix.org
When this happens the URL is assumed to be a file which then crashes the app once it's passed to the browser

# How
Create the URL with URLComponents which ensures it is [RFC 3986](https://www.ietf.org/rfc/rfc3986.txt) compliant. 

# Test Plan
Tested with multiple variations of URL and the one causing the issue. All work correctly.

### CODE REVIEW

File: packages/expo-modules-core/CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,7 @@
 ### 🐛 Bug fixes
 
 - Fix the `View cannot be cast to ViewGroup` exception on Android. ([#23264](https://github.com/expo/expo/pull/23264) by [@lukmccall](https://github.com/lukmccall))
+- Fix url converions that failed despite receiving a string that contained a valid URL. ([#23331](https://github.com/expo/expo/pull/23331) by [@alanhughes](https://github.com/alanjhughes))
```
Comment (tsapeta) [inline packages/expo-modules-core/CHANGELOG.md:12]:
```suggestion
- [iOS] Fix conversion to `URL` type that failed despite receiving a string that contained a valid URL. ([#23331](https://github.com/expo/expo/pull/23331) by [@alanhughes](https://github.com/alanjhughes))
```

**Review by tsapeta — COMMENTED** [review]
One related unit test fails. We can consider removing it though. What do you think?

**Comment by alanjhughes** [comment]
Yes, I think it's fine to remove it. The encoding get's handled correctly so we don't need that one

File: packages/expo-modules-core/ios/Swift/Arguments/Convertibles.swift [hunk 2]
```diff
@@ -17,11 +17,8 @@ extension URL: Convertible {
       throw Conversions.ConvertingException<URL>(value)
     }
 
-    // URLComponents parses and constructs URLs according to RFC 3986.
-    // For some unusual urls URL(string:) will fail incorrectly
-    if let components = URLComponents(string: value), let url = components.url {
-      // If it has no scheme, we assume it was the file path which needs to be recreated to be recognized as the file url.
-      return url.scheme != nil ? url : URL(fileURLWithPath: value)
+    if let url = convertToUrl(string: value) {
```
Comment (tsapeta) [inline packages/expo-modules-core/ios/Swift/Arguments/Convertibles.swift:20]:
```suggestion
    // First we try to create a URL without extra encoding, as it came.
    if let url = convertToUrl(string: value) {
```

**Review by tsapeta — APPROVED** [review]

**Pull request merged**